artichoke-backend/src/extn/core/env/env_object.rs

```diff
@@ -0,0 +1,295 @@
+use crate::convert::Convert;
+use crate::extn::core::error::{ArgumentError, RubyException};
+use std::marker::PhantomData;
+use std::mem;
+
+use crate::gc::MrbGarbageCollection;
+use crate::sys;
+use crate::value::Value;
+use crate::Artichoke;
+
+use super::backends::EnvBackend;
+use super::errors::EnvError;
+use mruby_sys::mrb_state;
+use std::collections::HashMap;
+
+pub trait RubyEnvNativeApi {
+    unsafe extern "C" fn get(mrb: *mut sys::mrb_state, _slf: sys::mrb_value) -> sys::mrb_value;
+    unsafe extern "C" fn set(mrb: *mut sys::mrb_state, _slf: sys::mrb_value) -> sys::mrb_value;
+    unsafe extern "C" fn env_to_h(mrb: *mut sys::mrb_state, _slf: sys::mrb_value)
+        -> sys::mrb_value;
+}
+
+pub struct Env<T: EnvBackend> {
+    phantom: PhantomData<T>,
+}
+
+impl<T: EnvBackend> Env<T> {
+    const TWO_STRINGS_ARGS_SPEC: &'static [u8] = b"SS!\0";
+
+    unsafe fn extract_two_string_args(interp: &Artichoke) -> (String, Option<String>) {
+        let mut key = <mem::MaybeUninit<sys::mrb_value>>::uninit();
+        let mut value = <mem::MaybeUninit<sys::mrb_value>>::uninit();
+        let mrb: *mut mrb_state = interp.borrow().mrb;
+        sys::mrb_get_args(
+            mrb,
+            Self::TWO_STRINGS_ARGS_SPEC.as_ptr() as *const i8,
+            key.as_mut_ptr(),
+            value.as_mut_ptr(),
+        );
+        let key = key.assume_init();
+        let value = value.assume_init();
+
+        let key_v = Value::new(interp, key);
+        let value_v = Value::new(interp, value);
+
+        if sys::mrb_sys_value_is_nil(value_v.inner()) {
+            (key_v.to_s(), None)
+        } else {
+            (key_v.to_s(), Some(value_v.to_s()))
+        }
+    }
+
+    // set_env
+    // This function may panic if key is empty, contains an ASCII equals sign '='
+    //      or the NUL character '\0', or when the value contains the NUL character.
+    fn validate_set_args(key: &str, value: &Option<String>) -> Result<(), EnvError> {
+        if key.find('=').is_some() || key.find('\0').is_some() {
+            return Err(EnvError::InvalidSetKey);
+        }
+
+        if value.is_some() && value.clone().unwrap().find('\0').is_some() {
+            return Err(EnvError::InvalidSetValue);
+        }
+
+        Ok(())
+    }
+    fn set_internal(key: &str, value: Option<&String>) {
+        T::set_value(key, value);
+    }
+
+    fn string_to_value(interp: &Artichoke, key: &str) -> Value {
+        let gc_was_enabled = interp.disable_gc();
+
+        let result = Value::convert(interp, key);
+
+        if gc_was_enabled {
+            interp.enable_gc();
+        }
+
+        result
+    }
+
+    fn hash_map_to_value(interp: &Artichoke, data: HashMap<String, String>) -> Value {
```

Comment:
Same as `string_to_value`, the GC management is not needed and all we're doing is calling the converter. This function is not needed.